aten/src/ATen/core/function_schema_inl.h

```diff
@@ -169,6 +192,86 @@ inline bool FunctionSchema::isBackwardCompatibleWith(
   return true;
 }
 
+inline bool FunctionSchema::isForwardCompatibleWith(
+    const FunctionSchema& old,
+    std::ostream* why_not) const {
+  if (!(name() == old.name() &&
+        overload_name() == old.overload_name()
+        // we are conservative on is_vararg and is_varret,
+        // since they are only used by internal operators
+        && is_vararg() == old.is_vararg() && is_varret() == old.is_varret() &&
+        returns().size() == old.returns().size() &&
+        arguments().size() >= old.arguments().size())) {
+    return false;
+  }
+
+  // find the start of out args in the old schema
+  int old_out_start_idx;
+  for (old_out_start_idx = old.arguments().size() - 1; old_out_start_idx > -1;
+       old_out_start_idx--) {
+    if (!old.arguments().at(old_out_start_idx).is_out()) {
+      break;
+    }
+  }
+
+  old_out_start_idx++;
+
+  // find the start of out args in the old schema
+  int new_out_start_idx;
+  for (new_out_start_idx = arguments().size() - 1; new_out_start_idx > -1;
+       new_out_start_idx--) {
+    if (!arguments().at(new_out_start_idx).is_out()) {
+      break;
+    }
+  }
+
+  new_out_start_idx++;
+
+  // make sure among the default args, they are forward compatible
+  for (size_t i = 0; i < old_out_start_idx; i++) {
+    if (!arguments().at(i).isForwardCompatibleWith(
+            old.arguments().at(i), why_not)) {
+      return false;
+    }
+  }
+
+  // Validate that all new arguments provided has a default value
+  for (size_t i = old_out_start_idx; i < new_out_start_idx; ++i) {
+    if (!arguments().at(i).default_value()) {
+      if (why_not) {
+        *why_not
+            << "Function schema not backward compatible since the new argument '"
+            << arguments().at(i).name() << "' of type "
+            << arguments().at(i).type()->str()
+            << " did not provide a default value.";
+      }
+      return false;
+    }
+
+    auto default_val = arguments().at(i).default_value().value();
+    if (default_val.isList() || default_val.isGenericDict()) {
+      if (why_not) {
+        *why_not
+            << "Function schema not forward compatible since the new argument '"
+            << arguments().at(i).name() << "' of type "
+            << arguments().at(i).type()->str() << " is a container type.";
```

Comment:
Add more information here about we don't support default values of container type.

Reply:
Sounds good.